test/CMakeLists.txt

```diff
@@ -1,100 +1,118 @@
-  channel_test_helper.hpp
-  test_helper.hpp
-)
-
-add_executable(future_test ${SOURCE} ${HEADERS})
-
-include_directories(${Boost_INCLUDE_DIRS})
-
-target_link_libraries(
-    future_test
-    ${Boost_LIBRARIES} 
-    ${CMAKE_THREAD_LIBS_INIT} 
-    ${CONAN_LIBS})
-
-set_property(TARGET future_test PROPERTY CXX_STANDARD 14)
-set_property(TARGET future_test PROPERTY CXX_STANDARD_REQUIRED ON)
-
-########## Serial queue ##########
-
-PROJECT(serial_queue_test)
-
-set(SOURCE
-    serial_queue_test.cpp
-)
-
-set(HEADERS
-)
-
-add_executable(serial_queue_test ${SOURCE} ${HEADERS})
-
-include_directories(${Boost_INCLUDE_DIRS})
-
-target_link_libraries(
-    serial_queue_test
-    ${Boost_LIBRARIES} 
-    ${CMAKE_THREAD_LIBS_INIT} 
-    ${CONAN_LIBS})
-
-set_property(TARGET serial_queue_test PROPERTY CXX_STANDARD 14)
-set_property(TARGET serial_queue_test PROPERTY CXX_STANDARD_REQUIRED ON)
-
-########## Copy On Write ##########
-
-PROJECT(cow_test)
-
-set(SOURCE
-    cow_test.cpp
-    main.cpp
-)
-
-set(HEADERS
-)
-
-add_executable(cow_test ${SOURCE} ${HEADERS})
-
-include_directories(${Boost_INCLUDE_DIRS})
-
-target_link_libraries(
-    cow_test
-    ${Boost_LIBRARIES} 
-    ${CMAKE_THREAD_LIBS_INIT} 
-    ${CONAN_LIBS})
-
-set_property(TARGET cow_test PROPERTY CXX_STANDARD 14)
-set_property(TARGET cow_test PROPERTY CXX_STANDARD_REQUIRED ON)
+add_executable( stlab.test.future.test
+                "${CMAKE_CURRENT_SOURCE_DIR}/channel_functor_tests.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/channel_join_tests.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/channel_merge_tests.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/channel_process_tests.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/channel_tests.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/channel_test_helper.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/channel_zip_tests.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/future_recover_tests.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/future_tests.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/future_then_tests.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/future_when_all_arguments_tests.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/future_when_all_range_tests.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/future_when_any_arguments_tests.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/future_when_any_range_tests.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/test_helper.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/tuple_algorithm_test.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/main.cpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/channel_test_helper.hpp"
+                "${CMAKE_CURRENT_SOURCE_DIR}/test_helper.hpp" )
+
+target_compile_options( stlab.test.future.test
+                        PUBLIC ${stlab_base_flags}
+                               $<$<CONFIG:DEBUG>:$<$<BOOL:${coverage}>:${stlab_coverage_flags}>${stlab_debug_flags}>
```

Comment:
Sure thing. This particular flavor of black magic is called a [generator expression](https://cmake.org/cmake/help/v3.2/manual/cmake-generator-expressions.7.html).

<background>
The syntax is admittedly horrendous, but it's the only mechanism CMake provides for supporting separate configuration and generation steps. For build tools like make and ninja, these steps are synonymous, so it's much ado about nothing, but for IDE tools like Visual Studios or Xcode, users are permitted to select whether to build in Debug or Release mode (generation step) after running CMake (configuration).
</background>

Would you like me to describe what they are/why they're useful in the comments? Or just what's being accomplished here?